src/OpenSearch/Client.php

```diff
@@ -1977,34 +1868,57 @@ public function extractArgument(array &$params, string $arg)
     }
 
     /**
-     * Sends a raw request to the cluster
-     * @return callable|array
-     * @throws NoNodesAvailableException
+     * Check if the client is running in async mode.
      */
-    public function request(string $method, string $uri, array $attributes = [])
+    public function isAsync(): bool
+    {
+        return $this->isAsync;
+    }
+
+    /**
+     * Set the client to run in async mode.
+     */
+    public function setAsync(bool $isAsync): static
+    {
+        $this->isAsync = $isAsync;
+        return $this;
+    }
+
+    /**
+     * Sends a raw request to the cluster.
+     *
+     * @throws \Psr\Http\Client\ClientExceptionInterface|\Exception
+     */
+    public function sendRawRequest(string $method, string $uri, array $params = [], ?string $body = null): Promise|ResponseInterface
     {
         $params = $attributes['params'] ?? [];
         $body = $attributes['body'] ?? null;
-        $options = $attributes['options'] ?? [];
 
-        $promise = $this->transport->performRequest($method, $uri, $params, $body, $options);
+        $request = $this->transport->createRequest($method, $uri, $params, $body);
 
-        return $this->transport->resultOrFuture($promise, $options);
+        if ($this->isAsync()) {
+            return $this->transport->sendAsyncRequest($request);
+        }
+        return $this->transport->sendRequest($request);
     }
 
     /**
-     * @return callable|array
+     * Perform the request.
+     *
+     * @throws \Psr\Http\Client\ClientExceptionInterface|\Exception
      */
-    private function performRequest(AbstractEndpoint $endpoint)
+    private function performRequest(EndpointInterface $endpoint): Promise|ResponseInterface
     {
-        $promise =  $this->transport->performRequest(
+        $request = $this->transport->createRequest(
             $endpoint->getMethod(),
             $endpoint->getURI(),
             $endpoint->getParams(),
             $endpoint->getBody(),
-            $endpoint->getOptions()
         );
-
-        return $this->transport->resultOrFuture($promise, $endpoint->getOptions());
+        if ($this->isAsync()) {
```

Comment:
this `isAsync` flag changing the return type makes it impossible to properly define the return type of methods of the SDK, as they will change return type depending on the configuration of the client, which would be a pain for DX.

Also, as discussed in https://github.com/opensearch-project/opensearch-php/pull/219#discussion_r1715828607, it looks like the existing version of opensearch-php does not actually support using the async mode for those methods (unless I missed a way to enable it), so maybe this switch should be removed entirely here, to allow keeping a well-typed SDK.